tfw.yml

```diff
@@ -48,6 +48,34 @@ builders:
   associate_public_ip_address: true
   subnet_id: "{{ user `travis_subnet_id` }}"
   vpc_id: "{{ user `travis_vpc_id` }}"
+- type: amazon-ebsvolume
+  access_key: "{{ user `aws_access_key` }}"
+  ebs_volumes:
+    - delete_on_termination: false
+      device_name: /dev/xvdx
+      tags:
+        Name: packer-docker-images
+        role: docker-images
+      volume_size: 80
+      volume_type: gp2
+  instance_type: t2.medium
+  region: us-east-1
+  run_tags:
+    Name: "packer-docker-images-generator"
+  secret_key: "{{ user `aws_secret_key` }}"
+  source_ami_filter:
+    filters:
+      name: "tfw*"
+      root-device-type: ebs
+      tag:role: "tfw"
+      virtualization-type: hvm
+    most_recent: true
+    owners:
+    - 341288657826
+  ssh_username: FIXME
+  subnet_id: "{{ user `travis_subnet_id` }}"
+  vpc_id: "{{ user `travis_vpc_id` }}"
+
```

Comment:
😮 whaaaaaat how does it do?  IIUC, this will mean that within a given `packer build` run, the `amazon-ebsvolume` builder will create an instance based on the most recent available AMI with a `role: tfw` tag and produce an EBS volume artifact, but otherwise the same provisioning steps are run (?).  Is that right? 😺

Reply:
That sounds _accurate_. I'm not sure it's necessarily _right_, though. :) To confirm I understand correctly, we want a `packer build` to:

1. generate an AMI (this is what [`type: amazon-ebs`](https://github.com/travis-ci/packer-templates/blob/master/tfw.yml#L30-L50) does, right?)
2. create an instance with the AMI generated in step 1, and:
- attach a brand new EBS volume to the instance
- run `start travis-worker` so that it pulls the docker images, then exit (I'm not sure how to make it shut down immediately after fetching images; TBD)
- shut down the instance (packer does this automatically as part of the [`amazon-ebsvolume`](
This builder builds EBS volumes by launching an EC2 instance from a source AMI, provisioning that running machine, and then destroying the source machine, keeping the volumes intact.) type)
- snapshot the volume and tag the snapshot

Any `terraform plan` run after that will get the latest snapshot with the `docker-images` tag (or whatever we want to call it), and the latest (or hardcoded) AMI.

This raises the following questions:
- If we're tying the EBS volume to an AMI (i.e. snapshots will only be generated when we generate a new AMI), is this approach actually superior to just baking the docker images into the AMI? 
- Will the EBS volume created by Packer be based on the AMI generated by the same `packer build`, or on the preceding one?

Reply:
In that case, I _think_ what we should do is define a separate packer template specifically for the steps you describe in preseeding the instance and saving off an EBS volume.  Something like a `tfw-worker-preseed.yml` that contains the `type: amazon-ebsvolume` provisioner and image lookup like you've defined here, but with provisioning steps that do what you described with starting `travis-worker` and pulling images.  I think this would mean that such a packer template would depend on the successful completion of:

- `tfw` `amazon-ebs` output
- `ci-amethyst` `docker` output
- `ci-garnet` `docker` output

Getting this all wired up via event based automation sounds rad to me, but as a first pass I'm also good with a daily Travis Cron job.

Have I sufficiently confused the topic? 😁